File: src/main/java/core/procedures/keywords/KeywordProc.java
package core.procedures.keywords;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.Keyword;

public final class KeywordProc extends AFn {

  public KeywordProc() {
    super(new FnArgsBuilder().min(1).max(1).mandatory(new Class[]{CharSequence.class}).build());
  }

  @Override
  public String getName() {
    return "keyword";
  }

  @Override
  public Keyword apply1(Object arg) {
    return arg == null ? null : Keyword.intern(arg.toString());
  }
}

File: src/main/kotlin/core/procedures/keywords/KeywordProc.kt
package core.procedures.keywords

import core.procedures.AFn
import core.scm.Keyword

class KeywordProc : AFn<CharSequence?, Keyword?>(name = "keyword", minArgs = 1, maxArgs = 1,
                                                 mandatoryArgsTypes = arrayOf<Class<*>>(CharSequence::class.java)) {

    override operator fun invoke(arg: CharSequence?) = arg?.let { Keyword.intern(arg.toString()) }
}
